controllers/imageclusterinstall_controller.go

```diff
@@ -537,41 +554,101 @@ func (r *ImageClusterInstallReconciler) SetupWithManager(mgr ctrl.Manager) error
+		return nil
+	}
+
+	return &dataImage
+}
+
+func (r *ImageClusterInstallReconciler) isInspectionEnabled(bmh *bmh_v1alpha1.BareMetalHost) bool {
+	if bmh == nil {
+		return false
+	}
+
+	if bmh.ObjectMeta.Annotations != nil && bmh.ObjectMeta.Annotations[inspectAnnotation] != "disabled" {
+		return true
+	}
+
+	return false
+}
+
+func (r *ImageClusterInstallReconciler) createBMHDataImage(ctx context.Context, bmh *bmh_v1alpha1.BareMetalHost, ici *v1alpha1.ImageClusterInstall, url string) error {
+	dataImage := r.getBmhDataImage(ctx, bmh)
+
+	if dataImage == nil {
+		r.Log.Infof("creating new dataImage for BareMetalHost (%s/%s)", bmh.Name, bmh.Namespace)
+
+		// Name and namespace must match the ones in BMH
+		dataImage = &bmh_v1alpha1.DataImage{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      bmh.Name,
+				Namespace: bmh.Namespace,
+			},
+			Spec: bmh_v1alpha1.DataImageSpec{
+				URL: url,
+			},
+		}
+		err := controllerutil.SetControllerReference(bmh, dataImage, r.Client.Scheme())
+		if err != nil {
+			return fmt.Errorf("failed to set controller reference for dataImage due to %w", err)
+		}
+
+		err = r.Create(ctx, dataImage)
+		if err != nil {
+			return fmt.Errorf("failed to create dataImage due to %w", err)
+		}
+	}
+
 	patch := client.MergeFrom(bmh.DeepCopy())
 
 	dirty := false
 	if !bmh.Spec.Online {
 		bmh.Spec.Online = true
+		r.Log.Infof("enabling BareMetalHost (%s/%s) Online", bmh.Name, bmh.Namespace)
 		dirty = true
 	}
-	if bmh.Spec.Image == nil {
-		bmh.Spec.Image = &bmh_v1alpha1.Image{}
-		dirty = true
-	}
-	if bmh.Spec.Image.URL != url {
-		bmh.Spec.Image.URL = url
-		dirty = true
+
+	if bmh.Status.Provisioning.State == bmh_v1alpha1.StateAvailable {
+		if !bmh.Spec.ExternallyProvisioned {
+			r.Log.Infof("Setting BareMetalHost (%s/%s) ExternallyProvisioned spec", bmh.Name, bmh.Namespace)
+			bmh.Spec.ExternallyProvisioned = true
+			dirty = true
+		}
 	}
-	liveIso := "live-iso"
-	if bmh.Spec.Image.DiskFormat == nil || *bmh.Spec.Image.DiskFormat != liveIso {
-		bmh.Spec.Image.DiskFormat = &liveIso
+
+	if bmh.Status.Provisioning.State == bmh_v1alpha1.StateAvailable && bmh.Status.PoweredOn &&
+		setAnnotaitonIfNotExists(&bmh.ObjectMeta, rebootAnnotation, "true") {
+		r.Log.Infof("Adding reboot annotations to BareMetalHost (%s/%s)", bmh.Name, bmh.Namespace)
 		dirty = true
 	}
 
 	if bmh.Spec.AutomatedCleaningMode != bmh_v1alpha1.CleaningModeDisabled {
 		bmh.Spec.AutomatedCleaningMode = bmh_v1alpha1.CleaningModeDisabled
+		r.Log.Infof("Disable automated cleaning mode for BareMetalHost (%s/%s)", bmh.Name, bmh.Namespace)
 		dirty = true
 	}
 
-	if bmh.Status.Provisioning.State == bmh_v1alpha1.StateProvisioned {
+	// boot from disk
+	if bmh.Status.Provisioning.State == bmh_v1alpha1.StateExternallyProvisioned && bmh.Status.PoweredOn {
 		if setAnnotaitonIfNotExists(&bmh.ObjectMeta, detachedAnnotation, detachedAnnotationValue) {
+			r.Log.Infof("Adding detached annotations to BareMetalHost (%s/%s)", bmh.Name, bmh.Namespace)
```

Comment:
Wait, won't this prevent ironic from performing the reboot?

Reply:
It should actually, when we started the installation we should not allow reboots

Reply:
So let's move this to when the installation is complete

Reply:
Why? Isn't taht the case we want to reboot the host?

Reply:
We actually want bmh not to be rebooted when installation finished , i believe we should set it right after we started installation but setting it after checkStatus should be ok too